docs(specs): add Slots API spec (#16529)
microsoft/fluentui: specs/Slots.md

# Slots API specification

## Slots definition

> Fluent UI components almost always contain sub parts and these pre-defined layouts can be configured via Slots API. Slots are named areas in a component that can receive props and provide styling and layout for them.

Reviewer: ```suggestion
Fluent UI components are often composed internally with different subparts to help with layouts. These pre-defined layouts can be configured via the Slots API. Slots are named areas in a component that can receive props for styling and layout.
```

For example, `Button` component contains `content`, `loader` and `icon` slots in its layout:

```jsx
// ⚠️ not a real JSX/DOM markup
<slots.root>
  <slots.loader />
  <slots.children />
  <slots.icon />
</slots.root>
```

> Slots let define the complex layouts required by many components according to accessibility and design requirements automatically. Slots can be configured through props, which lets the caller pass in a variety of inputs for a given slot.

Reviewer: ```suggestion
Slots define the complex layouts required by many components according to accessibility and design requirements automatically. Slots can be configured through props, which lets the caller pass in a variety of inputs for a given slot.
```

## Slots components

Each slot is represented by a React component or a primitive, for example: `slots.loader` can be a `Loader` component and `slots.icon` can be a `span`:

```jsx
// ⚠️ not a real JSX/DOM markup
<>
  {/* renders a <button /> element */}
  <slots.root>
    {/* renders a <Loader /> element */}
    <slots.loader />
    {/* renders a <span /> element */}
    <slots.icon />
  </slots.root>
</>
```

> A default set of components is manually configured per each component, but is optional, by default slots are represented by `div` elements.

Reviewer: Not quite sure what you're trying to say here, do you mean that is is possible to manually configure each slot ?

Author: Yes, may be you can suggest a better phrase there:
- components have slots
- each slot by default is `div` element
- this can be overridden

Reviewer: ```suggestion
Slots are rendered as `div` elements by default. However, all rendering aspects of a slot can be manually configured and overridden in the ways described below
```

Author: ```suggestion
Slots are rendered as `div` elements by default. However, all rendering aspects of a slot can be manually configured and overridden.
```

I cut a part of phrase as it needs decisions from an upcoming RFC.

## Slots usage

> There are several forms of values that can be provided, but all of them share one common thing - they will produce a proper layout inside a component.

Reviewer: ```suggestion
Slots can be configured with different types and shapes of values, but result should not change - slots will produce a correct layout inside a component.
```

### An object as a value

> If you will pass an object to component's props we will handle it as props for a slot, this behavior can be considered as declaring a JSX element with a javascript object. In an example below, props will be passed to an `icon` slot:

Reviewer: I think the first part is unnecessary, given the analogy and code sample

```suggestion
This behavior can be considered as declaring a JSX element with a javascript object. In an example below, props will be passed to an `icon` slot:
```

```jsx
<Button icon={{ children: <FooIcon />, className: 'an-awesome-slot', id: '#button-icon' }} />
```

```html
<!-- 💡 Simplified DOM output -->
<button class="ms-Button">
  <!-- 👇 An additional class and id have been added to markup -->
  <span class="ms-Button-icon an-awesome-slot" id="#button-icon">
    <!-- 👇 An icon will be rendered inside a slot as it is passed as children -->
    <span class="ms-Icon"><svg /></span>
  </span>
</button>
```

> Any props what are valid for a component can be passed to a slot, i.e. `data-*` attributes and event handlers:

Reviewer: ```suggestion
Any props which are valid for a component can be passed to a slot, i.e. `data-*` attributes and event handlers:
```

```jsx
<Button
  icon={{
    children: <FooIcon />,
    'data-test-id': 'button-foo-icon',
    onClick: () => {
      console.log('A click on an icon slot!');
    },
  }}
/>
```

### Primitives as a value

We consider JSX elements as primitive values that will be passed to a meaningful prop, by default to `children` (but this can be configured in the component's implementation):

```jsx
<>
  <Button icon={{ children: <FooIcon /> }} />
  {/* 💡 will produce the same JSX/HTML markup, as will be expanded to { children: <FooIcon /> } */}
  <Button icon={<FooIcon />} />
</>
```

Such usage of slots is called shorthands, similarly to [CSS shorthand properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties). The same behavior is applicable strings and numbers:

```jsx
<>
  <Button icon={{ children: 'text instead of icon' }} />
  {/* 💡 will produce the same JSX/HTML markup */}
  <Button icon="text instead of icon" />
</>
```

> To disable slot rendering you can use falsy (`null`, `false`) values:

Reviewer: The one exception I can think of here would be `0` or empty string, should they be allowed to stop a slot from rendering ? 

Or could we limit this to `null` `undefined` and `false` ? which seem to make the most sense for me in the context of slot rendering

Author: > The one exception I can think of here would be 0 or empty string, should they be allowed to stop a slot from rendering ?

```suggestion
To disable slot rendering you can use `null` & `false` values:
```

I think this will be clearer statement. As `0` and `''` should be handled in the same way as strings and numbers.

P.S. `undefined` in React fallbacks to a default value (https://codesandbox.io/s/sleepy-haze-9g75g), so I think that we should support only `false` & `null`.

```jsx
// 💡 A loader slot will be hidden
<>
  <Button loading loader={null} />
  <Button loading loader={false} />
</>
```

`true` will render a slot without any content:

```jsx
// 💡 A loader slot will be renderer without any content
<Button loading loader={true} />
```

```html
<!-- 💡 Simplified DOM output -->
<button class="ms-Button">
  <span class="ms-Button-loader"></span>
</button>
```

### Renders props via `children` function

> Slots can be deeply customized via `children` function that behave similarly to [Render Props pattern](https://reactjs.org/docs/render-props.html) in React:

Reviewer: I think that this is the actual goal of the `children` render func right ? to not have to throw away the original result

```suggestion
Slots can be deeply customized without ignoring the original content by using `children` as a function that behave similarly to [Render Props pattern](https://reactjs.org/docs/render-props.html) in React:
```

  <Button icon={<FooIcon />} />
</>
```

Such usage of slots is called shorthands, similarly to [CSS shorthand properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties). The same behavior is applicable strings and numbers:

```jsx
<>
  <Button icon={{ children: 'text instead of icon' }} />
  {/* 💡 will produce the same JSX/HTML markup */}
  <Button icon="text instead of icon" />
</>
```

To disable slot rendering you can use `null` as a value:

```jsx
// 💡 A loader slot will be hidden
<Button loading loader={null} />
```

### Renders props via `children` function

Slots can be deeply customized without ignoring the original content by using `children` as a function that behave similarly to [Render Props pattern](https://reactjs.org/docs/render-props.html) in React:

```jsx
<Button
  icon={{
    // 💡 "Component" is an original element type for slot, for example, it can be a "span"
    //    "props" are defaults that are provided to a slot by a host component, for example, may contain "onClick"
    //    handler
    children: (Component, props) => (
      <div id="icon-wrapper">

>         {/* 💡 "props" can be modified there to override component's behavior */}

Reviewer 2: ```suggestion
        {/* 💡 "props" can be modified here to override component's behavior */}
```

        <Component {...props} id="icon-slot">
          <FooIcon />
        </Component>
      </div>
    ),
  }}
/>
```

```html
<!-- 💡 Simplified DOM output -->
<button class="ms-Button">
  <!-- 👇 An additional element was added to markup -->
  <div id="icon-wrapper">
    <span class="ms-Button-icon" id="icon-slot">
      <span class="ms-Icon"><svg /></span>
    </span>
  </div>
</button>
```

> ⚠️An important note is that `children` function replaces the whole slot, not its contents.

Reviewer 2: ```suggestion
⚠️An important note is that `children` function replaces the whole slot, not its contents.

## Slot for collections

Besides slots for single (scalar) values, v0/northstar also supports slots for collections (array of items in `Menu` component).

Converged components will **not** support slots for collections but will rather use Children API in these cases.
```

# Slots API specification

Fluent UI components almost always contain sub parts and these pre-defined layouts can be configured via Slots API. Slots are named areas in a component that can receive props and provide styling and layout for them.

## Slots definition

For example, `Button` component contains `content`, `loader` and `icon` slots in its layout:

```jsx
<root>
  <loader />
  <content />
  <icon />
</root>
```

## Slots usage

Slots can be configured through props, which lets the caller pass in a variety of inputs for a given slot. There are several forms of values that can be provided, but all of them share one common thing - they will produce a proper JSX layout inside a component.

### An object as a value

> If you will pass an object to component props we will handle it as props for a slot. In an example below, props will be passed to an `icon` slot:

Reviewer: I think we shold also mention that any attributes that are possible to use for standard JSX are fair game in shorthand

Reviewer: it's a non-standard prop that we use quite frequently

Author: @ling1726 good suggestion 👍 I added an explicit section about this, is it clear now? Do you have any other suggestions?

Reviewer: The extra code sample makes this a lot clearer, thanks.

```jsx

> <Button icon={{ children: <FooIcon />, id: '#button-icon' }} />

Reviewer 3: **q:** I'm bit confused about the API TBH. 

This example will render following?

```diff
<root>
  <loader />
  <content />
+  icon slot render as -> <FooIcon id="button-icon" />
</root>
```

or like this?

```diff
<root>
  <loader />
  <content />
+ <icon id="button-icon">
+     <FooIcon />
+</icon>
</root>
```

Reviewer 4: I'm not 100% sure but I think it will be the former. @layershifter am I correct here?

Reviewer 4: Actually, I think it's always (or by default) within a `span` element. See [react-utils/src/compose/getSlots.ts#L33](https://github.com/microsoft/fluentui/blob/master/packages/react-utils/src/compose/getSlots.ts#L33).

Author: @kubkon @Hotell `id` will go to a slot while `icon` will be rendered inside. I agree that this without an output it's confusing, so I added a snippet with rendered HTML. Is it cleaner now?

```

### Primitives as a value

We consider JSX elements, strings and numbers as primitive values that will be passed to a meaningful prop, usually to `children`:

```jsx
<>
  <Button icon={<FooIcon />} />
  {/* 💡 has identical effect to the previous one */}

>   <Button icon={{ children: <FooIcon /> }} />

Reviewer: Is the second example syntactic sugar for the first ?

Reviewer: or should they work completely differently under the hood ?

Reviewer: No need for tech implem details, but we should be clear if one is an alias for the other

Reviewer 4: I think the first example is just a shorthand for the more explicit second example according to [react-examples/src/compose/getSlots.ts#L33](https://github.com/microsoft/fluentui/blob/master/packages/react-utils/src/compose/getSlots.ts#L33). Ofc please do correct me if I'm wrong!

Reviewer 3: So I looked briefly at current impl of `getSlots` and this statement is misleading (correct me if I'm wrong). 

From JSX transpilation perspective:

`<Button icon={<FooIcon />} />` -> `<Button icon={React.createElement(FooIcon)} />`  ->`<Button icon={{ type: FooIcon, children: null }} />`

how can that be identical to `<Button icon={{ children: <FooIcon /> }} />` ? Is the implementation gonna change ?

Reviewer 4: OK, so I just checked and you're right guys, implementation wise, the two are different:
1. generates 

```
Button props:
icon: <FooIcon />
```

2. generates

```
Button props:
icon: { children: <FooIcon /> }
```

When rendered in the DOM, however, they do come down to the same set of elements:

```html
<button ...>
    <span class="... icon-"
         <!-- whatever FooIcon is rendered as -->
    </span>
</button>
```

Author: @kubkon @ling1726 @Hotell thanks for feedback 👍 

I updated this block to (text around it was also updated):

```jsx
<>
  <Button icon={{ children: <FooIcon /> }} />
  {/* 💡 will produce the same JSX/HTML markup, as will be expanded to { children: <FooIcon /> } */}
  <Button icon={<FooIcon />} />
</>
```

(hope that now it has more sense)

We have a special handling for JSX elements, strings and number, so they will be expanded to `{ children: VALUE }`. In code it's done in `resolveShorthands()`:

https://github.com/microsoft/fluentui/blob/661d1d82d79e606a7fa77ed1125ffc6528325ec9/packages/react-utilities/src/compose/resolveShorthandProps.tsx#L22-L25

</>
```

> To disable slot rendering you can use falsy values:

Reviewer 3: **q:** falsy or null ? I'd recommend to explicit/follow js language practices -> thus allow `null` only

Author: I modified this section to be more explicit: "To disable slot rendering you can use falsy (`null`, `false`) values:" and also mentioned `true` behavior below.

We are already handling `false` & `null` in this way I think that we should include `false` as this behavior can be unclear for customers:

```jsx
<>
  <Button loader={null} /> // hides slot
  <Button loader={false} /> // will not hide a slot if will not include "false"
</>
```

Reviewer: I resolved a separate comment referring to this issue after seeing this one again.

@layershifter Does this mean that only `null, false` are supported ? In this case I would refrain from using the term `falsy` since it relates to a larger set of possible values (0, empty string)
